### PR TITLE
SUS-5293 Add maintenance tasks to queue

### PR DESCRIPTION
**Pull request by fraszczakszymon**
Example:
```
SERVER_ID=177 php maintenanceTaskScheduler.php --id=165,177 --active=1 "maintenanceScriptClass" --options --args
```

### CODE REVIEW

File: maintenance/MaintenanceTask.php [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+class MaintenanceTask extends \Wikia\Tasks\Tasks\BaseTask {
+	public function run( $script ) {
+		return shell_exec( $script );
```
Comment (macbre) [inline maintenance/MaintenanceTask.php]: Let's validate `$script` - is its path relative to app's root?  If so, let's check if `file_exists( $IP. $script)` (with `$script` sanitized and `../` being removed :) )

File: maintenance/maintenanceTaskScheduler.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+require_once( dirname( __FILE__ ) . "/Maintenance.php" );
+
+class MaintenanceTaskScheduler extends Maintenance {
+	public function __construct() {
+		parent::__construct();
+
+		$this->mDescription = "Adds task execution to the queue";
+		$this->addArg( "script", "Script that is going to be executed", true );
+		$this->addOption( "active", "Run on wikis which are active at least <param> days" );
+		$this->addOption( "id", "Wiki IDs (separated with comma)" );
```
Comment (macbre) [inline maintenance/maintenanceTaskScheduler.php]: `--cluster` option of `run_maintenance` is used by cron jobs as well

**Review by macbre — APPROVED** [review]

File: maintenance/maintenanceTaskScheduler.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		$idsParam = [];
+		if ( $this->getOption( "id" ) !== null ) {
+			$idsParam = array_map( "intval", explode( ",", $this->getOption( "id" ) ) );
+		}
+
+		$this->output( "Scheduling ${script}\n" );
+
+		$db = wfGetDB( DB_SLAVE, [], $wgExternalSharedDB );
+
+		// TODO: Handle 300k wikis
```
Comment (macbre) [inline maintenance/maintenanceTaskScheduler.php:34]: Let's use MediaWiki's database access layer and simply iterate over `$res->fetchObject()` and yield rows.

File: maintenance/maintenanceTaskScheduler.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		$this->getWikis( $idsParam, $active, $cluster )
+			->runLoop( $db, function ( $data, $row ) use ( $script ) {
+				$this->scheduleTask( $row->city_id, $script );
+			} );
+	}
+
+	private function getWikis( array $idsParam = [], int $active = null, string $cluster = null ) {
+		$sql = ( new \WikiaSQL() )
+			->SELECT( "city_id" )
+			->FROM( "city_list" );
```
Comment (macbre) [inline maintenance/maintenanceTaskScheduler.php]: Please add `city_public = 1` condition - we only want to process active wikis and skip closed ones.

File: maintenance/maintenanceTaskScheduler.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+		return $sql;
+	}
+
+	private function scheduleTask( int $wikiId, string $script ) {
+		$task = new \Wikia\Tasks\Tasks\MaintenanceTask();
+		$task->wikiId( $wikiId );
+		$task->call( "run", $script );
+
+		$task_id = $task->queue();
```
Comment (mszabo-wikia) [inline maintenance/maintenanceTaskScheduler.php]: let's call queue only once and pass many wiki IDs—this way we create only 1 rabbit channel and connection for publishing

File: maintenance/maintenanceTaskScheduler.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+			->FROM( "city_list" );
+
+		if ( count( $idsParam ) > 0 ) {
+			$sql->WHERE( "city_id" )
+				->IN( $idsParam );
+		}
+
+		if ( $active !== null ) {
+			$timestampSql = "city_last_timestamp BETWEEN TIMESTAMP( DATE_SUB(CURDATE(), INTERVAL %d DAY) ) AND NOW()";
+			$sql->AND_( sprintf( $timestampSql, $active ) );
```
Comment (mszabo-wikia) [inline maintenance/maintenanceTaskScheduler.php]: should be escaped

**Review by mszabo-wikia — CHANGES_REQUESTED** [review]

File: maintenance/maintenanceTaskScheduler.php (hunk 3 above)
Comment (JamesHawking) [reply maintenance/maintenanceTaskScheduler.php:34]: It does on line #34.

File: maintenance/maintenanceTaskScheduler.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+			->runLoop( $db, function ( $data, $row ) use ( $script ) {
+				$this->scheduleTask( $row->city_id, $script );
+			} );
+	}
+
+	private function getWikis( array $idsParam = [], int $active = null, string $cluster = null ) {
+		$sql = ( new \WikiaSQL() )
+			->SELECT( "city_id" )
+			->FROM( "city_list" )
+			->WHERE("city_public")->GREATER_THAN(WikiFactory::CLOSE_ACTION);
```
Comment (sqreek) [inline maintenance/maintenanceTaskScheduler.php]: `equal 1` ?

Comment (JamesHawking) [reply maintenance/maintenanceTaskScheduler.php]: Macbre suggested to use this instead of 1.

File: extensions/wikia/WikiFactory/WikiFactory.php [hunk 8]
```diff
@@ -58,10 +58,11 @@ class WikiFactory {
...
 
-	const db            = "wikicities"; // @see $wgExternalSharedDB
-	const DOMAINCACHE   = "/tmp/wikifactory/domains.ser";
-	const CACHEDIR      = "/tmp/wikifactory/wikis";
-	const WIKIA_TOP_DOMAIN = '.wikia.com';
+	const db            	= "wikicities"; // @see $wgExternalSharedDB
+	const DOMAINCACHE   	= "/tmp/wikifactory/domains.ser";
+	const CACHEDIR      	= "/tmp/wikifactory/wikis";
+	const WIKIA_TOP_DOMAIN 	= '.wikia.com';
+	const PUBLIC_WIKI      	= 1;
```
Comment (sqreek) [inline extensions/wikia/WikiFactory/WikiFactory.php]: please put this const near other city_public const (LINE 45-46)

File: maintenance/maintenanceTaskScheduler.php [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	public function execute() {
+		global $wgExternalSharedDB;
+
+		$script = $this->getArg( 0 );
+		$active = $this->getOption( "active" );
+		$cluster = $this->getOption( "cluster" );
+
+		$idsParam = [];
+		if ( $this->getOption( "id" ) !== null ) {
+			$idsParam = array_map( "intval", explode( ",", $this->getOption( "id" ) ) );
```
Comment (mszabo-wikia) [inline maintenance/maintenanceTaskScheduler.php]: `str_getcsv` is better instead of explode

File: maintenance/maintenanceTaskScheduler.php [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	private function scheduleTask( array $wikisId, string $script ) {
+		$task = new \Wikia\Tasks\Tasks\MaintenanceTask();
+		$task->wikiId( $wikisId );
+		$task->call( "run", $script );
+
+		$task_id = $task->queue();
+
+		$this->output( "Scheduled ${task_id}\n" );
```
Comment (mszabo-wikia) [inline maintenance/maintenanceTaskScheduler.php]: this will be an array, not a single ID

**Review by mszabo-wikia — CHANGES_REQUESTED** [review]

File: extensions/wikia/WikiFactory/WikiFactory.php [hunk 11]
```diff
@@ -44,6 +44,7 @@ class WikiFactory {
 	# close Wiki
 	const HIDE_ACTION 			= -1;
 	const CLOSE_ACTION 			= 0;
+	const PUBLIC_WIKI      		= 1;
```
Comment (sqreek) [inline extensions/wikia/WikiFactory/WikiFactory.php]: formatting

**Review by mszabo-wikia — APPROVED** [review]

File: maintenance/maintenanceTaskScheduler.php [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+		if ( $cluster !== null ) {
+			$sql->AND_( 'city_cluster' )
+				->EQUAL_TO( $cluster );
+		}
+
+		return $sql;
+	}
+
+	private function scheduleTask( array $wikisId, string $script ) {
```
Comment (sqreek) [inline maintenance/maintenanceTaskScheduler.php]: 's

File: maintenance/maintenanceClasses.php [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+$wgAutoloadClasses['CloseMyAccountMaintenance'] = __DIR__ . '../extensions/wikia/CloseMyAccount/maintenance/CloseMyAccountMaintenance.php';
+$wgAutoloadClasses['ResetWeeklyUserContributionsCount'] = __DIR__ . 'wikia/cronjobs/resetWeeklyUserContributionsCount.php';
+//$wgAutoloadClasses[''] = __DIR__ . 'wikia/discussion-alerts.php';
+$wgAutoloadClasses['UpdateThumbnail'] = __DIR__ . 'wikia/VideoHandlers/updateThumbnail.php';
+$wgAutoloadClasses['TaskManagerArchive'] = __DIR__ . '../extensions/wikia/TaskManager/maintenance/archive.php';
+//$wgAutoloadClasses[''] = __DIR__ . 'wikia/SiteWideMessagesMaintenance.php';
+//$wgAutoloadClasses[''] = __DIR__ . '../extensions/wikia/WallNotifications/maintenance.php';
+//$wgAutoloadClasses[''] = __DIR__ . '../extensions/TorBlock/loadExitNodes.php';
+$wgAutoloadClasses['UploadStashCleanup'] = __DIR__ . 'cleanupUploadStash.php';
+$wgAutoloadClasses['SendConfirmationReminder'] = __DIR__ . '../extensions/wikia/AuthPages/maintenance/sendConfirmationReminder.php';
```
Comment (macbre) [inline maintenance/maintenanceClasses.php]: I don't really like this autoloader. It's yet another place to remember to add your script to. Can't we use `wfShellExec` to run the script?

Additionally, not all script are class-based. Some are just inline code in the PHP file.

**Pull request merged**